### PR TITLE
feat(ptu): accrue flat cost for PTU deployments declared in config.yaml

### PR DESCRIPTION
**Pull request by yucheng-berri**
## TLDR

Problem this solves:

- A PTU deployment in config.yaml accrues no flat cost
- The rollup only reads the database model table
- Its traffic is billed per token instead

How it solves it:

- The rollup also reads config-declared deployments off the router
- Registering one zeroes its pricing, since capacity is prepaid
- Both halves apply the same rules from one module

## User Flow

Before: a team on reserved capacity is billed for its tokens and never for the capacity

1. An admin declares a provisioned-throughput deployment in `config.yaml` with `team_id`, `ptu_count`, `cost_per_ptu_per_hour` and `ptu_effective_from`, then starts the proxy with `LITELLM_ENABLE_PTU_COST_ATTRIBUTION=True`
2. Someone on that team sends POST https://litellm-domain/v1/chat/completions and gets a normal answer
3. https://litellm-domain/ui/?page=logs shows that request charged at the provider's per-token rate
4. The nightly attribution job runs at 00:15 UTC and writes nothing for that deployment
5. https://litellm-domain/ui/?page=usage shows the team's token spend and no reserved-capacity cost, so the hourly charge the provider is invoicing appears nowhere

After: the same team is billed for the capacity it reserved, and not again for the traffic that capacity serves

1. The admin declares the same deployment and starts the same proxy
2. Someone on that team sends the same POST https://litellm-domain/v1/chat/completions and gets the same answer
3. https://litellm-domain/ui/?page=logs shows that request at $0, because the reservation already paid for it
4. The nightly job runs and attributes the reservation to the team
5. https://litellm-domain/ui/?page=usage shows the team's reserved-capacity cost for the day, charted separately from request cost

An admin who wants the deployment billed per token instead removes `ptu_count` and `cost_per_ptu_per_hour`, exactly as on the API path

## Linear ticket

Resolves LIT-5809

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have added meaningful tests
- [x] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [ ] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## Screenshots / Proof of Fix

Real Postgres, real Gemini traffic, a proxy booted from the config below on each side, no mocks. Gemini stands in for Azure because PTU capacity cannot be provisioned here; the fields the feature reads are provider-agnostic, and what is being proven is the cost path. The only thing not exercised is the cron trigger itself, since the job has no on-demand entry point, so each side calls the same scheduled function directly against its own database.

```yaml
model_list:
  - model_name: gpt-4o-ptu
    litellm_params:
      model: gemini/gemini-2.5-flash
      api_key: os.environ/GEMINI_API_KEY
      input_cost_per_token: 0.000005
      output_cost_per_token: 0.000015
    model_info:
      team_id: ptu-team-0001
      ptu_count: 100
      cost_per_ptu_per_hour: 0.02
      ptu_effective_from: "2026-01-01T00:00:00Z"

general_settings:
  master_key: sk-lit5809
```

```bash
export LITELLM_ENABLE_PTU_COST_ATTRIBUTION=True
python -m litellm.proxy.proxy_cli --config ptu.yaml --port <port>
KEY=$(curl -s -X POST http://127.0.0.1:<port>/key/generate -H "Authorization: Bearer sk-lit5809" \
  -H 'Content-Type: application/json' -d '{"team_id":"ptu-team-0001","models":["gpt-4o-ptu"]}' | jq -r .key)
```

#### a real request through the config.yaml PTU deployment

1. Send it

```bash
curl -s -X POST http://127.0.0.1:4811/v1/chat/completions -H "Authorization: Bearer $KEY" \
  -H 'Content-Type: application/json' \
  -d '{"model":"gpt-4o-ptu","messages":[{"role":"user","content":"Say hi."}],"max_tokens":25}'
```

```
tokens: 4 / 21
```

2. The deployment is priced at what the operator declared

```
router pricing: input=5e-06 output=1.5e-05
```

#### the nightly attribution job

1. Run it for that day

```
rollup: RollupResult(day=2026-08-18, models_processed=0, rows_written=0, rows_failed=0, lapsed=())
flat-cost rows: []
```

2. Read the team's day back out of Postgres

```
                             api_key                              | model_group |  spend   | prompt_tokens | completion_tokens | ptu_flat_cost
------------------------------------------------------------------+-------------+----------+---------------+-------------------+---------------
 291e0dcf9ff6ecfbb84b5306a9cd6fb2575ca8a2fa992d859ad81593d58ba670 | gpt-4o-ptu  | 0.000335 |             4 |                21 |             0
```

The tokens are billed at 4 x 5e-06 plus 21 x 1.5e-05, and the reservation the provider is invoicing hourly is recorded nowhere

#### a real request through the config.yaml PTU deployment

1. Send the same request

```bash
curl -s -X POST http://127.0.0.1:4812/v1/chat/completions -H "Authorization: Bearer $KEY" \
  -H 'Content-Type: application/json' \
  -d '{"model":"gpt-4o-ptu","messages":[{"role":"user","content":"Say hi."}],"max_tokens":25}'
```

```
tokens: 4 / 21
```

2. The deployment carries no per-token rate

```
router pricing: input=0.0 output=0.0
```

#### the nightly attribution job

1. Run it for that day

```
rollup: RollupResult(day=2026-08-18, models_processed=1, rows_written=1, rows_failed=0, lapsed=())
flat-cost rows: [('ptu-team-0001', 'gpt-4o-ptu', 48.0)]
```

2. Read the team's day back out of Postgres

```
                             api_key                              | model_group | spend | prompt_tokens | completion_tokens | ptu_flat_cost
------------------------------------------------------------------+-------------+-------+---------------+-------------------+---------------
 8d808bffd673ab86d6f0e3a2addf8f5933ed7f26fd5a6b9d2666ed5aa52c0864 | gpt-4o-ptu  |     0 |             4 |                21 |             0
 __ptu_flat_cost__                                                | gpt-4o-ptu  |     0 |             0 |                 0 |            48
```

100 units at $0.02 an hour over a full day is $48, the same request records 4 and 21 real tokens at no per-token charge, and the deployment id is unchanged from the before run, so cooldowns, deployment budgets and existing spend rows all still key on it

Re-run at the current tip, exit code checked and the log grepped for a traceback, after an earlier capture turned out to be a partially completed run

## Type

🆕 New Feature

## Changes

The rollup reads the router's deployments alongside the table, keeping the ones no database row owns and dropping the per-request credential clones, which carry `original_model_id` and would otherwise bill one reservation once per distinct caller key.

Registering such a deployment zeroes its pricing. The rules for what accrues and what that zeroes live in `litellm/litellm_core_utils/ptu_pricing.py`, which the write endpoints now import as well, because a deployment the rollup declines to charge and the router prices at zero would serve its traffic for free. `tiered_pricing` is emptied rather than zeroed, since its tiers outrank the rates beside them; the search context table is written zeroed because an absent one means the provider's default rather than free; and any further rate the deployment declares is zeroed alongside the standing set.

The prune is bounded to the deployments a run scanned, but only for a run that priced a config-declared deployment, so a database-only proxy sweeps exactly as it does today.

The nightly catch-up shares that loader, so a config-declared reservation is priced for its elapsed days too. Verified live: a reservation started five days ago writes five days at $240, one started two hundred days ago writes 91 days at $4,368, and a second run of either writes nothing further.

## Caveats (if any)

- A zeroed deployment wins QualityRouter's cost tiebreak where an unset rate lost it
- Attribution needs every replica booted from the same config.yaml
- Outside its window a PTU deployment bills nothing, matching the database path
- Enabling this prices up to 91 elapsed days on the first nightly run

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

---

<a href="https://app.devin.ai/review/berriai/litellm/pull/37556" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: litellm/proxy/spend_tracking/ptu_flat_cost_rollup.py [hunk 1]
```diff
@@ -358,10 +312,71 @@ async def _upsert_charge_with_retry(
...
+    scanned_ids: frozenset[str]
+    config_sourced: bool
+
+
+def _running_router() -> object | None:
+    """The proxy's router, or None outside a running proxy.
+
+    Read out of ``sys.modules`` rather than imported, so a rollup driven from a test or a
+    script does not pull the whole proxy server in behind it.
+    """
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/spend_tracking/ptu_flat_cost_rollup.py:335]: Correct, and inherent to config.yaml being per-pod. Divergent files break serving before billing. Added a caveat: every replica must boot from the same file.

File: litellm/router.py [hunk 2]
```diff
@@ -7695,7 +7696,16 @@ def _create_deployment(
         - None: If the deployment is not active for the current environment (if 'supported_environments' is set in litellm_params)
         """
         try:
-            litellm_params: Final[LiteLLM_Params] = LiteLLM_Params(**_litellm_params)
+            zeroed_pricing: Final = (
+                zeroed_ptu_pricing(_model_info, _litellm_params) if _model_info.get("db_model") is not True else None
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/router.py:7700]: Pre-existing: #36829 zeroes on carrying PTU config, not window state, so the database path already does this. Gating here alone would split the paths.

File: litellm/litellm_core_utils/ptu_pricing.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    are checked first only because they are a few dict reads, while the flag can resolve
+    through a configured secret manager, and this runs for every deployment registered.
+
+    Any further rate the deployment itself declares is zeroed alongside the standing set,
+    since one left standing bills the traffic the reserved capacity already paid for.
+    """
+    if ptu_terms(model_info) is None:
+        return None
+    if not is_ptu_cost_attribution_enabled():
+        return None
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/litellm_core_utils/ptu_pricing.py:138]: Agreed, and as you note this is the database path's existing behaviour from #36829. Changing it here alone would split the two paths.

**Comment by yucheng-berri** [comment]
@greptileai please review the current head c7ca8f4dcf88610838d1e4005e76f4f93c332350. Divergent replica configs break serving before billing; documented as a caveat, and unfixable without persisting config models.

File: litellm/router.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/router.py:7700]: #36829 exercised a closed window live and states it: "A closed window alerts; it does not resume per-token billing."

File: litellm/litellm_core_utils/ptu_pricing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/litellm_core_utils/ptu_pricing.py:138]: Confirmed deliberate: #36829 demonstrated a closed window live and documented it as "A closed window alerts; it does not resume per-token billing."

**Review by yuneng-berri — APPROVED** [review]
